core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenTaskSourceFactory.java

```diff
@@ -189,7 +213,8 @@ private SplitAssigner createSplitAssigner(
                     sourcePartitioningScheme,
                     outputDataSizeEstimates,
                     fragment,
-                    getFaultTolerantExecutionTargetTaskInputSize(session).toBytes());
+                    // number of tasks are bounded for hash distribution so adaptive task sizing is automatic
+                    partitioning.equals(SCALED_WRITER_HASH_DISTRIBUTION) ? writerTaskInitTargetInputSizeInBytes : nonWriterTaskInitTargetInputSizeInBytes);
```

Comment:
We should have adaptive sizing for `HashDistributionSplitAssigner` where ware dealing with writer fragment.
In that case we splitting single hash partitions which can be huge into multiple tasks. And logic should be the same. Initially crated tasks should get less data and task created later should get more data.

I guess you can parametrize `HashDistributionSplitAssigner` with same set of parameters as `ArbitraryDistributionSplitAssigner` but make the "growth" ones optional. That is
```
            OptionalInt adaptiveTaskSizingStepSize,
            OptionalDouble adaptiveTaskSizingGrowthFactor,
            long initTargetPartitionSizeInBytes,
            OptionalLong maxPartitionSizeInBytes,
```

Then the Optional parameters will be set and used only for writer case.

Or maybe you can skip Optionals and just use initial and max same for non-writer case.